File: app/schemas/lead.py
from datetime import date, datetime
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field


LeadStatus = Literal[
    "new",
    "to_call",
    "contacted",
    "trial_scheduled",
    "trial_done",
    "converted",
    "lost",
]

SportExperience = Literal["debutant", "sportif", "confirme", "crossfit"]


class LeadBase(BaseModel):
    prenom: str = Field(..., min_length=1, max_length=255)
    nom: str = Field(..., min_length=1, max_length=255)
    email: EmailStr
    telephone: Optional[str] = Field(default=None, max_length=50)
    experience: Optional[SportExperience] = None
    message: Optional[str] = None

    source: Optional[str] = None
    utm_source: Optional[str] = None
    utm_medium: Optional[str] = None
    utm_campaign: Optional[str] = None
    utm_content: Optional[str] = None
    utm_term: Optional[str] = None
    referrer: Optional[str] = None
    landing_url: Optional[str] = None


class LeadWaitlistCreate(LeadBase):
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None


class LeadUpdate(BaseModel):
    status: Optional[LeadStatus] = None
    lost_reason: Optional[str] = None
    notes: Optional[str] = None
    next_action_date: Optional[date] = None


class LeadInDBBase(LeadBase):
    id: UUID
    status: LeadStatus
    lost_reason: Optional[str] = None
    notes: Optional[str] = None
    next_action_date: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class LeadPublic(LeadInDBBase):
    pass


class LeadEventPublic(BaseModel):
    id: UUID
    event_type: str
    event_note: Optional[str] = None
    event_at: datetime

    class Config:
        from_attributes = True


class LeadWithEvents(LeadInDBBase):
    events: List[LeadEventPublic] = []


class LeadListResponse(BaseModel):
    items: List[LeadPublic]
    total: int
    page: int
    page_size: int
    pages: int

File: backend/src/types/lead.ts
export type LeadStatus =
  | "new"
  | "to_call"
  | "contacted"
  | "trial_scheduled"
  | "trial_done"
  | "converted"
  | "lost";

export type SportExperience =
  | "debutant"
  | "sportif"
  | "confirme"
  | "crossfit";

export interface LeadEvent {
  id: string;
  event_type: string;
  event_note: string | null;
  event_at: string;
}

export interface Lead {
  id: string;
  prenom: string;
  nom: string;
  email: string;
  telephone: string | null;
  experience: SportExperience | null;
  message: string | null;
  source: string | null;
  utm_source: string | null;
  utm_medium: string | null;
  utm_campaign: string | null;
  utm_content: string | null;
  utm_term: string | null;
  referrer: string | null;
  landing_url: string | null;
  status: LeadStatus;
  lost_reason: string | null;
  notes: string | null;
  next_action_date: string | null;
  created_at: string;
  updated_at: string;
}

export interface LeadWithEvents extends Lead {
  events: LeadEvent[];
}

export interface LeadListResponse {
  items: Lead[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface LeadUpdate {
  status?: LeadStatus;
  lost_reason?: string | null;
  notes?: string | null;
  next_action_date?: string | null;
}

export interface DashboardStats {
  total: number;
  new_today: number;
  converted: number;
  to_call_today: number;
  by_status: Partial<Record<LeadStatus, number>>;
  by_source: Record<string, number>;
}
